### PR TITLE
Automatically log non-successful RPCs

### PR DESCRIPTION
**Pull request by codingllama**
Log errored RPCs using unary and stream interceptors.

Example:

```
2023-07-26T18:20:15-03:00 DEBUG [GRPC]      finished call grpc.code:AlreadyExists grpc.component:server grpc.error:rpc error: code = AlreadyExists desc = asset tag already registered grpc.method:CreateDevice grpc.method_type:unary grpc.service:teleport.devicetrust.v1.DeviceTrustService grpc.start_time:2023-07-26T18:20:15-03:00 grpc.time_ms:0.62 peer.address:127.0.0.1:51341 protocol:grpc v2@v2.0.0-rc.3/logger.go:33
```

### CODE REVIEW

**Comment by codingllama** [comment]
Admittedly this is a bit silly, as I could easily write a custom interceptor instead of pulling from `grpc-gateway`. Let me know if you prefer that.

File: lib/auth/middleware.go [hunk 1]
```diff
@@ -465,46 +467,63 @@ func (a *Middleware) withAuthenticatedUserStreamInterceptor(srv interface{}, ser
 	return handler(srv, &authenticatedStream{ctx: ctx, ServerStream: serverStream})
 }
 
+// loggingInterceptorOpts is used with [logging.UnaryServerInterceptor] and
+// [logging.StreamServerInterceptor].
+var loggingInterceptorOpts = []logging.Option{
+	logging.WithDecider(func(_ string, err error) logging.Decision {
```
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/auth/middleware.go:473]: I noticed that too, but I want to backport this and I wasn't sure about updating the deps in release branches. I can do a quick follow up, master only, after this PR.

File: lib/auth/middleware.go [hunk 2]
```diff
@@ -465,46 +467,63 @@ func (a *Middleware) withAuthenticatedUserStreamInterceptor(srv interface{}, ser
...
+	if a.GRPCMetrics != nil {
+		is = append(is, om.UnaryServerInterceptor(a.GRPCMetrics))
+	}
+
+	is = append(is,
 		utils.GRPCServerUnaryErrorInterceptor,
 		a.Limiter.UnaryServerInterceptorWithCustomRate(getCustomRate),
 		a.withAuthenticatedUserUnaryInterceptor,
 	)
+	return utils.ChainUnaryServerInterceptors(is[0], is[1:]...)
```
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/auth/middleware.go:502]: Done and done.

**Review by rosstimothy — COMMENTED** [review]
> 2023-07-26T18:20:15-03:00 INFO [GRPC]

Should we limit these new logs to DEBUG level? I'm worried that if we start filling up customer logs with every RPC failure it will cause much confusion. Some of the failures may be expected and ultimately just add noise.

Edit:

A few examples where a failed RPC doesn't indicate a failure:

- Creating a role: `2023-07-27T10:38:44-04:00 INFO [GRPC]      finished call grpc.code:NotFound grpc.component:server grpc.error:rpc error: code = NotFound desc = role new_role_name is not found grpc.method:GetRole grpc.method_type:unary grpc.service:proto.AuthService grpc.start_time:2023-07-27T10:38:44-04:00 grpc.time_ms:1.299 peer.address:127.0.0.1:53705 protocol:grpc v2@v2.0.0-rc.3/logger.go:33`
- Creating an auth connector: `2023-07-27T10:33:43-04:00 INFO [GRPC]      finished call grpc.code:NotFound grpc.component:server grpc.error:rpc error: code = NotFound desc = github connector "github" is not configured grpc.method:GetGithubConnector grpc.method_type:unary grpc.service:proto.AuthService grpc.start_time:2023-07-27T10:33:43-04:00 grpc.time_ms:0.573 peer.address:127.0.0.1:53645 protocol:grpc v2@v2.0.0-rc.3/logger.go:33`
- Agents checking if there are any network restrictions:  `2023-07-27T10:42:32-04:00 INFO [GRPC]      finished call grpc.code:NotFound grpc.component:server grpc.error:rpc error: code = NotFound desc = key /restrictions/network is not found grpc.method:GetNetworkRestrictions grpc.method_type:unary grpc.service:proto.AuthService grpc.start_time:2023-07-27T10:42:32-04:00 grpc.time_ms:0.513 peer.address:127.0.0.1:53825 protocol:grpc v2@v2.0.0-rc.3/logger.go:33`
- Agents are stopped: `2023-07-27T10:43:46-04:00 ERRO [GRPC]      finished call grpc.code:Unknown grpc.component:server grpc.error:rpc error: code = Unknown desc = inventory control stream recv failed: rpc error: code = Canceled desc = context canceled grpc.method:InventoryControlStream grpc.method_type:bidi_stream grpc.service:proto.AuthService grpc.start_time:2023-07-27T10:42:32-04:00 grpc.time_ms:74307.76 peer.address:127.0.0.1:53824 protocol:grpc v2@v2.0.0-rc.3/logger.go:37`

File: lib/auth/middleware.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/auth/middleware.go]: Changed to a custom interceptor and moved all logs to Debug-level.

**Comment by codingllama** [comment]
PTAL?

We can silence some codes if we like, like `codes.NotFound` and maybe even `codes.InvalidArgument`, but I'd rather keep `codes.Unknown` - it's precisely the latter that motived this PR.

**Comment by zmb3** [comment]
Related: #17186

File: api/utils/grpc/logging.go
[the remark this answers is not in the document]
Comment (codingllama) [reply api/utils/grpc/logging.go:1]: Done.

**Comment by codingllama** [comment]
I wanted the error logging specifically for tracking unexpected behavior on Cloud. I can log just the RPCs I'm concerned about ATM, but I do think that would be a waste.

> Do we need to worry about the contents of an error landing in the logs

Based on the few examples you provided we could filter out NotFound errors, but in general erroring with sensitive info is a bad idea - and one we can't solve in this PR.

**Comment by codingllama** [comment]
After chatting a bit more yesterday, it seems that this has little chance to land. I've moved on to logging in the specific RPCs I care about. If you'd like me to bring this back to life, give me a nudge here.

**Pull request closed without merging**